### PR TITLE
fix(go)!: Fallback to git-tags instead of github-tags

### LINKED ISSUE #17923: Renovate is incorrectly using the github-tags host type for the 'Go' datasource
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

v32.202.0 (latest)

### If you're self-hosting Renovate, select which platform you are using.

Bitbucket Server

### If you're self-hosting Renovate, tell us what version of the platform you run.

v7.21.4

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

When using a custom import path, Renovate will fail to fetch the dependency because it assumes it's GitHub, which in our case isn't.

https://pkg.go.dev/cmd/go#hdr-Remote_import_paths

Example:

We have a simple server `code.example.com` which redirects Go imports to `git.example.com` by responding with the correct meta tag.

```html
<meta name="go-import" content="code.example.com/repo/project git ssh://git@git.example.com/repo/project">
```

For whatever reason, Renovate then decides this is a `github-tags` host type and creates dependency lookup warnings.

I don't see any obvious reason why it would do this given the code.

https://github.com/renovatebot/renovate/blob/4168b34e7b37a195b406c93984d94069ca60cf77/lib/modules/datasource/go/base.ts#L76

### Relevant debug logs

<details><summary>Logs</summary>

```
renovate-27729360-g8f7w renovate {"name":"renovate","hostname":"renovate-27729360-g8f7w","pid":3,"level":30,"logContext":"lZTCHZzVfwYRMOxFnblaE","repository":"repo/project1","baseBranch":"main","stats":{"managers":{"gomod":{"fileCount":1,"depCount":4}},"total":{"fileCount":1,"depCount":4}},"msg":"Dependency extraction complete","time":"2022-09-21T14:55:46.242Z","v":0}
renovate-27729360-g8f7w renovate {"name":"renovate","hostname":"renovate-27729360-g8f7w","pid":3,"level":20,"logContext":"lZTCHZzVfwYRMOxFnblaE","repository":"repo/project1","host":"code.example.com","msg":"No concurrency limits","time":"2022-09-21T14:55:46.278Z","v":0}
renovate-27729360-g8f7w renovate {"name":"renovate","hostname":"renovate-27729360-g8f7w","pid":3,"level":20,"logContext":"lZTCHZzVfwYRMOxFnblaE","repository":"repo/project1","goModule":"code.example.com/repo/project2","goImportURL":"ssh://**redacted**@git.example.com/repo/project2","msg":"Go lookup import url","time":"2022-09-21T14:55:46.349Z","v":0}
renovate-27729360-g8f7w renovate {"name":"renovate","hostname":"renovate-27729360-g8f7w","pid":3,"level":20,"logContext":"lZTCHZzVfwYRMOxFnblaE","repository":"repo/project1","host":"git.example.com","msg":"No concurrency limits","time":"2022-09-21T14:55:46.354Z","v":0}
renovate-27729360-g8f7w renovate {"name":"renovate","hostname":"renovate-27729360-g8f7w","pid":3,"level":20,"logContext":"lZTCHZzVfwYRMOxFnblaE","repository":"repo/project1","err":{"name":"UnsupportedProtocolError","code":"ERR_UNSUPPORTED_PROTOCOL","message":"Unsupported protocol \"ssh:\"","stack":"UnsupportedProtocolError: Unsupported protocol \"ssh:\"\n    at normalizeArguments (/usr/src/app/node_modules/got/dist/source/core/index.js:521:23)\n    at got (/usr/src/app/node_modules/got/dist/source/create.js:112:39)\n    at gotRoutine (/usr/src/app/node_modules/renovate/lib/util/http/index.ts:71:27)\n    at queueTask (/usr/src/app/node_modules/renovate/lib/util/http/index.ts:160:16)\n    at GithubHttp.request (/usr/src/app/node_modules/renovate/lib/util/http/index.ts:167:45)\n    at GithubHttp.request (/usr/src/app/node_modules/renovate/lib/util/http/github.ts:316:34)\n    at GithubHttp.requestJson (/usr/src/app/node_modules/renovate/lib/util/http/index.ts:219:28)\n    at GithubHttp.getJson (/usr/src/app/node_modules/renovate/lib/util/http/index.ts:230:17)\n    at GithubTagsDatasource.getReleases (/usr/src/app/node_modules/renovate/lib/modules/datasource/github-tags/index.ts:91:23)\n    at GoDirectDatasource.getReleases (/usr/src/app/node_modules/renovate/lib/modules/datasource/go/releases-direct.ts:59:33)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at /usr/src/app/node_modules/renovate/lib/util/cache/package/decorator.ts:78:20\n    at /usr/src/app/node_modules/renovate/lib/util/cache/package/decorator.ts:78:20\n    at fetchReleases (/usr/src/app/node_modules/renovate/lib/modules/datasource/index.ts:300:13)\n    at getPkgReleases (/usr/src/app/node_modules/renovate/lib/modules/datasource/index.ts:352:7)\n    at lookupUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:80:32)\n    at fetchDepUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/fetch.ts:58:15)\n    at /usr/src/app/node_modules/p-map/index.js:57:22","options":{"headers":{"user-agent":"RenovateBot/32.201.2 (https://github.com/renovatebot/renovate)","accept":"application/vnd.github.v3+json"},"url":"ssh://git.example.com/api/v3/repos/repo/project2/tags?per_page=100","hostType":"github-tags","username":"","password":"","method":"GET","http2":false}},"msg":"Unknown GitHub error","time":"2022-09-21T14:55:46.356Z","v":0}
renovate-27729360-g8f7w renovate {"name":"renovate","hostname":"renovate-27729360-g8f7w","pid":3,"level":20,"logContext":"lZTCHZzVfwYRMOxFnblaE","repository":"repo/project1","datasource":"go","packageName":"code.example.com/repo/project2","errCode":"ERR_UNSUPPORTED_PROTOCOL","msg":"Datasource connection error","time":"2022-09-21T14:55:46.356Z","v":0}
renovate-27729360-g8f7w renovate {"name":"renovate","hostname":"renovate-27729360-g8f7w","pid":3,"level":20,"logContext":"lZTCHZzVfwYRMOxFnblaE","repository":"repo/project1","dependency":"code.example.com/repo/project2","packageFile":"go.mod","msg":"Failed to look up dependency code.example.com/repo/project2","time":"2022-09-21T14:55:46.356Z","v":0}
```

</details>

### Have you created a minimal reproduction repository?

No reproduction, but I have linked to a public repo where it occurs

### PR DESCRIPTION
**Pull request by uhthomas**
## Changes

Unknown imports are not guaranteed to be backed by GitHub, making it an unsafe fallback option. A generic Git variant should be better.

## Context

Fixes #17923.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Needs testing on a public repo

**Comment by uhthomas** [comment]
@rarkins So one problem with this is some situations will fallback to `git-tags` which don't need to.

For instance, a GitHub Enterprise server under a domain other than `github.com` will use `git-tags` rather than `github-tags`. I don't know how else to fix this, or if this behaviour is acceptable?

**Comment by viceice** [comment]
we've recently added as detect platform function. maybe that can be reused to make a more proper detect of the right datasource?

**Comment by uhthomas** [comment]
> we've recently added as detect platform function. maybe that can be reused to make a more proper detect of the right datasource?

That sounds like it could be helpful. Platform detection with this change could cover both cases?

**Review by rarkins — COMMENTED** [review]
Needs some public repo testing, including both:
- GitHub repo case, which should keep working, and
- non-GitHub case (which this fixes, wasn't working before)

File: lib/modules/datasource/go/base.ts [hunk 1]
```diff
@@ -207,7 +208,7 @@ export class BaseGoDatasource {
       .join('/');
 
     return {
-      datasource: GithubTagsDatasource.id,
+      datasource: GitTagsDatasource.id,
       registryUrl: `${parsedUrl.protocol}//${parsedUrl.host}`,
```
Comment (rarkins) [inline lib/modules/datasource/go/base.ts:212]: git-tags does not support registryUrls, so I don't think the change is as simple as this. However I thought this had been tested?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/go/base.ts (hunk 1 above)
Comment (uhthomas) [reply lib/modules/datasource/go/base.ts:202]: The field `registryUrls` is required in order to comply with the `DataSource` interface. You are right that I was missing some things though.

**Comment by rarkins** [comment]
When I run this branch against the reproduction repo, I still get failures:

```
DEBUG: Go lookup import url (repository=renovate-reproductions/17923)
       "goModule": "go-srht.uranus.workers.dev/~uhthomas/renovate17923",
       "goImportURL": "ssh://**redacted**@git.sr.ht/~uhthomas/renovate17923"
DEBUG: Datasource unknown error (repository=renovate-reproductions/17923)
       "datasource": "go",
       "packageName": "go-srht.uranus.workers.dev/~uhthomas/renovate17923",
       "err": {
         "task": {
           "commands": [
             "ls-remote",
             {
               "protocol": "ssh:",
               "slashes": true,
               "auth": "git",
               "host": "git.sr.ht",
               "port": null,
               "hostname": "git.sr.ht",
               "hash": null,
               "search": null,
               "query": null,
               "pathname": "/~uhthomas/renovate17923",
               "path": "/~uhthomas/renovate17923",
               "href": "ssh://**redacted**@git.sr.ht/~uhthomas/renovate17923"
             }
           ],
           "format": "utf-8",
           "parser": "[function]"
         },
         "message": "fatal: ']' does not appear to be a git repository\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n",
         "stack": "Error: fatal: ']' does not appear to be a git repository\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n\n    at Object.action (/Users/rhys/src/github/renovatebot/renovate/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)\n    at PluginStore.exec (/Users/rhys/src/github/renovatebot/renovate/node_modules/simple-git/src/lib/plugins/plugin-store.ts:29:29)\n    at /Users/rhys/src/github/renovatebot/renovate/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:127:42\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/Users/rhys/src/github/renovatebot/renovate/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:14)\n    at GitExecutorChain.<anonymous> (/Users/rhys/src/github/renovatebot/renovate/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:100:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/Users/rhys/src/github/renovatebot/renovate/node_modules/simple-git/dist/cjs/index.js:55:24)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)"
       }
DEBUG: Failed to look up go dependency go-srht.uranus.workers.dev/~uhthomas/renovate17923 (repository=renovate-reproductions/17923, packageFile=go.mod, dependency=go-srht.uranus.workers.dev/~uhthomas/renovate17923)
```

**Comment by rarkins** [comment]
Debugging it, the `source.packageName` field is an object instead of a string:

image

I have fixed this (see my commit) but it now fails further on. This command does not work:

```
❯ git ls-remote ssh://git@git.sr.ht/~uhthomas/renovate17923
The authenticity of host 'git.sr.ht (173.195.146.142)' can't be established.
ED25519 key fingerprint is SHA256:WXXNZu0YyoE3KBl5qh4GsnF1vR0NeEPYJAiPME+P09g.
This key is not known by any other names
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added 'git.sr.ht' (ED25519) to the list of known hosts.
git@git.sr.ht: Permission denied (publickey,keyboard-interactive).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

**Comment by rarkins** [comment]
When viewed with ignore white space changes:

image

**Comment by rarkins** [comment]
I validated this by removing the gopkg.in logic and finding that it could still look it up using git tags.

**Comment by rarkins** [comment]
I'm not really sure this should count as a breaking change btw, but it's targeting v33

**Review by viceice — APPROVED** [review]
🤷‍♂️ let's take it as breaking

**Comment by viceice** [comment]
missing coverage

**Comment by uhthomas** [comment]
> Debugging it, the `source.packageName` field is an object instead of a string:
> 
> image
> 
> I have fixed this (see my commit) but it now fails further on. This command does not work:
> 
> ```
> ❯ git ls-remote ssh://git@git.sr.ht/~uhthomas/renovate17923
> The authenticity of host 'git.sr.ht (173.195.146.142)' can't be established.
> ED25519 key fingerprint is SHA256:WXXNZu0YyoE3KBl5qh4GsnF1vR0NeEPYJAiPME+P09g.
> This key is not known by any other names
> Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
> Warning: Permanently added 'git.sr.ht' (ED25519) to the list of known hosts.
> git@git.sr.ht: Permission denied (publickey,keyboard-interactive).
> fatal: Could not read from remote repository.
> 
> Please make sure you have the correct access rights
> and the repository exists.
> ```

FWIW I think this behaviour is fine and acceptable. Go modules using SSH will need to explicitly declare a known_hosts file with `ssh-keyscan` or configure SSH to not use strict host key checking.

```
Host *
    StrictHostKeyChecking no
```

**Review by rarkins — APPROVED** [review]

**Pull request merged**